File: 543.DiameterofBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(node) {
    if (node) {
        return 1 + Math.max(maxDepth(node.left), maxDepth(node.right));
    }
    return 0;
};

var maxDepth = function(node) {
    const stack = [node];
    let count = 1;
    let cursor = 0;
    let depth = 0;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            cursor++;
            if (cursor === count) {
                count = stack.length;
                cursor = 0;
                depth++;
            }
        }
    }
    return depth;   
}
// The diameter of a binary tree can be either
// 1. it's the diameter of the left sub tree if not passing through root node [fallback to recursive]
// 2. it's the diameter of the right sub tree if not passing through root node [fallback to recursive]
// 3. the sum of the max length of both of the root's child nodes [maxth depth of a binary tree, either BFS or recursive max + 1]
var diameterOfBinaryTree = function(root) {
    if (root) {
        return Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right), maxDepth(root.left) + maxDepth(root.right));
    }
    return 0;
};

File: Golang/Tree/543.Diameter-of-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	ret := 0
	var helper func(node *TreeNode) int
	helper = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		left := helper(node.Left)
		right := helper(node.Right)
		ret = max(ret, max(1+left+right, max(left, right)))
		return 1 + max(left, right)
	}
	helper(root)
	return ret - 1
}
